### PR TITLE
Introduce `causal_depthwise_conv` in the Conformer convolution module.

### PR DESCRIPTION
**Pull request by kauterry**
**What does this PR do? Please describe:**

Pad the input sequence entirely on the left for the Conv1d making it causal.

**Does your PR introduce any breaking changes? If yes, please list them:**
List of all backwards-incompatible changes.

**Check list:**
- [ ] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [ ] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**?
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

### CODE REVIEW

File: src/fairseq2/models/conformer/convolution.py [hunk 1]
```diff
@@ -14,6 +15,49 @@
...
+            bias=bias,
+            padding_mode=padding_mode,
+            device=device,
+            dtype=dtype,
+        )
+        # Pad the last dimension entirely on the left to make the convolution causal.
+        self.pad = (kernel_size - 1, 0)
+
+    def forward(self, x: Tensor) -> Tensor:
+        x = F.pad(x, self.pad)
```
Comment (cbalioglu) [inline src/fairseq2/models/conformer/convolution.py:57]: Do we use this `CausalConv1d` anywhere else beyond a `ConformerConvolution`? If not, I don't think that it is necessary to have a module for this functionality. We can achieve the same within `ConformerConvolution`'s forward method like:

```python
def forward(...):
   ...

   if self.depthwise_conv_is_causal:
       seqs = F.pad(seqs, (self.depthwise_conv.kernel_size[0] - 1, 0))
   
   seqs = self.depthwise_conv(seqs)
```

Comment (cbalioglu) [reply src/fairseq2/models/conformer/convolution.py:57]: Also, do we have any literature reference for this padding technique?

File: src/fairseq2/models/conformer/convolution.py [hunk 2]
```diff
@@ -42,6 +87,8 @@ def __init__(
             The dimensionality of the model.
         :param depthwise_kernel_size:
             The kernel size of the depthwise convolution.
+        :param depthwise_conv_is_causal:
```
Comment (cbalioglu) [inline src/fairseq2/models/conformer/convolution.py:90]: Nit: I would prefer this to be named `causal_depthwise_conv` which is a bit more aligned with the naming convention we use in other places in the code base.

File: src/fairseq2/models/conformer/convolution.py [hunk 3]
```diff
@@ -42,6 +87,8 @@ def __init__(
...
+            Is the depthwise convolution causal or not.
```
Comment (cbalioglu) [inline src/fairseq2/models/conformer/convolution.py:91]: Nit: For boolean parameters, we typically use the convention "If ``True``, uses a causal depthwise convolution." or even better if we have a literature reference "If ``True``, uses a causal depthwise convolution as described in :cite:t`xyz`".

File: src/fairseq2/models/conformer/convolution.py [hunk 4]
```diff
@@ -66,18 +113,32 @@ def __init__(
...
-            depthwise_kernel_size,
-            # We preserve the sequence length regardless of the kernel size.
-            padding="same",
-            # We want to perform depthwise convolution.
-            groups=model_dim,
-            bias=False,
-            device=device,
-            dtype=dtype,
-        )
+        if depthwise_conv_is_causal:
```
Comment (cbalioglu) [inline src/fairseq2/models/conformer/convolution.py:116]: Again, as I mentioned above. To me it looks like the only difference between these two modules is the extra pad statement in `CausalConv1d`. If that is true, we can certainly move it to `ConformerConvolution` and avoid this extra convolution module.

File: src/fairseq2/models/s2t_transformer/builder.py [hunk 5]
```diff
@@ -90,6 +90,9 @@ class S2TTransformerConfig:
     depthwise_conv_kernel_size: int
     """The kernel size of depthwise convolutions in Conformer blocks."""
 
+    depthwise_conv_is_causal: bool
```
Comment (cbalioglu) [inline src/fairseq2/models/s2t_transformer/builder.py:93]: S2T Transformer architecture has never supported causal depthwise convolution. We should not expose this as a configurable parameter in S2T Transformer.

File: src/fairseq2/models/w2vbert/builder.py [hunk 6]
```diff
@@ -43,6 +43,7 @@ def _encoder_600m() -> Wav2Vec2EncoderConfig:
         layer_drop_p=0.0,
         norm_order=TransformerNormOrder.POST,
         depthwise_conv_kernel_size=31,
+        depthwise_conv_is_causal=False,
```
Comment (cbalioglu) [inline src/fairseq2/models/w2vbert/builder.py:46]: Since the default is already `False` and this parameter is not relevant for w2v-BERT please remove this explicit argument pass.

File: src/fairseq2/models/w2vbert/builder.py [hunk 7]
```diff
@@ -75,6 +76,7 @@ def _encoder_300m() -> Wav2Vec2EncoderConfig:
...
```
Comment (cbalioglu) [inline src/fairseq2/models/w2vbert/builder.py:79]: Same as above.

File: src/fairseq2/models/wav2vec2/builder.py [hunk 8]
```diff
@@ -139,6 +139,9 @@ class Wav2Vec2EncoderConfig:
     depthwise_conv_kernel_size: int
     """The kernel size of depthwise convolutions in Conformer blocks."""
 
+    depthwise_conv_is_causal: bool
```
Comment (cbalioglu) [inline src/fairseq2/models/wav2vec2/builder.py:142]: Same parameter name and docstr (one line below) suggestion as in conformer/convolution.py.

File: src/fairseq2/models/conformer/convolution.py [hunk 9]
```diff
@@ -6,6 +6,7 @@
 
 from typing import Optional
 
+import torch.nn.functional as F
```
Comment (cbalioglu) [inline src/fairseq2/models/conformer/convolution.py:9]: Recently I have started using explicit imports (and updated the code base to follow that convention), so in the rest of the code base we use `from torch.nn.functional import pad`. I personally prefer this style since it gives you immediately the overview of that features/ops we use at the top of the file.

**Review by cbalioglu — COMMENTED** [review]
Mostly looks good. I have one feedback about the existence of the `CausalConv1d` module though. Let me know what you think. Otherwise, it is pretty close to be merged.

File: src/fairseq2/models/conformer/convolution.py (hunk 1 above)
Comment (kauterry) [reply src/fairseq2/models/conformer/convolution.py:57]: I believe it's used in WaveNet and similar architectures: https://www.deepmind.com/blog/wavenet-a-generative-model-for-raw-audio

It's a pretty common idea, so I wondered if we should have it as a module for anyone to leverage.

File: src/fairseq2/models/w2vbert/builder.py (hunk 6 above)
Comment (cbalioglu) [reply src/fairseq2/models/w2vbert/builder.py:46]: Nevermind. Ignore my comment. This is `Wav2Vec2EncoderConfig` so this parameter is required to be passed.

File: src/fairseq2/models/conformer/convolution.py (hunk 1 above)
Comment (kauterry) [reply src/fairseq2/models/conformer/convolution.py:57]: Alrighty.

File: src/fairseq2/models/conformer/convolution.py [hunk 10]
```diff
@@ -42,6 +46,9 @@ def __init__(
             The dimensionality of the model.
         :param depthwise_kernel_size:
             The kernel size of the depthwise convolution.
+        :param causal_depthwise_conv:
+            If True, uses a causal depthwise convolution similar to that described in
+            Section 2.1 of :cite:t:`https://doi.org/10.48550/arxiv.2006.11477`.
```
Comment (cbalioglu) [inline src/fairseq2/models/conformer/convolution.py:51]: This is wav2vec 2.0 paper right? I don't see a Section 2.1 or a reference to causality there. Are you sure that this is the right reference. Also I forgot to mention it in the first review, but you should also put the bibtex of the paper to bibliography.bib at the root of the repository.

Comment (kauterry) [reply src/fairseq2/models/conformer/convolution.py:51]: Oh no, wrong link sorry!

**Review by cbalioglu — APPROVED** [review]
